subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/BeanNode.java

```diff
@@ -19,32 +19,132 @@
 import com.google.common.base.Function;
 import com.google.common.base.Preconditions;
 import com.google.common.collect.Iterators;
+import org.gradle.api.Named;
 
 import javax.annotation.Nullable;
+import java.util.HashSet;
 import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
 
-public class BeanNode extends AbstractPropertyNode<BeanNode> {
-    private final Object bean;
+public abstract class BeanNode extends AbstractPropertyNode<BeanNode> {
 
-    public BeanNode(@Nullable String propertyName, Object bean) {
+    public static BeanNode create(@Nullable String propertyName, Object bean) {
+        if (bean instanceof Map<?, ?>) {
+            return new MapBeanNode(propertyName, (Map<?, ?>) bean);
+        }
+        if (bean instanceof Iterable<?>) {
+            return new IterableBeanNode(propertyName, (Iterable<?>) bean);
+        }
+        return new SimpleBeanNode(propertyName, bean);
```

Comment:
Do we already unpack `Callable`s when we get here?

Reply:
[We do not](https://github.com/gradle/gradle/blob/889fc7f3d6dced867f7d56e5af4a8425bfcf071b/subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/DefaultPropertyWalker.java#L83-L100). Do you think we should? Is this maybe part of supporting `Provider`s?

Reply:
Could be, yes. I think `Map<String, Callable<Map<String, Iterable<Callable<Named>>>>>` should be supported, no? I don't think we should bother with lazy keys for maps, though.

Reply:
We don't even support resolving a `Callable<NestedBean>` kind of thing, yet. That all falls into a similar kind of scheme: call `DeferredUtil.unpack` on nested properties, right?
I think this should be a separate PR, too.

Reply:
Thinking some more about it, I don't think we should unpack values for maps or iterables.
We added a new feature here. If someone wants to be lazy, they can use `Provider`s. If they don't, a simple `Map` will suffice.

If we try to unpack things here, we may break use-cases like tracking the implementations of `Closure`s or `Callable`s.

Moreover, I don't see the the compelling use-case for having `Callable`s in `Map` values. The getter annotated with `@Nested` for the map can create the map lazily anyway.

Reply:
FTR, we discussed to address this separately.